src/kinship/hierarchy-creator.ts

```diff
@@ -0,0 +1,349 @@
+    return indiIds.flatMap(id => this.indiIdToFamAsSpouseNodes(id, parentNode, childNodeType));
+  }
+
+  private indiIdToFamAsSpouseNodes(indiId: string, parentNode: TreeNode, childNodeType: LinkType): TreeNode[] {
+    if (this.isChildNodeTypeForbidden(childNodeType, parentNode)) return [];
+    const famsIds = this.data.getIndi(indiId).getFamiliesAsSpouse();
+    if (!famsIds.length) {
+      const node = this.idToNode(EntryId.indi(indiId), parentNode, childNodeType);
+      return node ? [node] : [];
+    }
+
+    const famsNodes: TreeNode[] = famsIds.map(id => {
+      return {
+        id,
+        indi: {id: indiId},
+        family: {id},
+        parentNode,
+        linkFromParentType: childNodeType,
+        childNodes: null,
+        linkStubs: null,
+        duplicateOf: this.queuedNodesById.get(id),
+      };
+    });
+    famsNodes.forEach((node, i) => {
+      if (i !== 0) node.primaryMarriage = famsNodes[0];
+      if (node.duplicateOf) node.duplicateOf.duplicated = true;
+      else this.queuedNodesById.set(node.id, node);
+    });
+    return famsNodes;
+  }
+
+  private famAsSpouseIdsToNodes(famsIds: string[], parentNode: TreeNode, childNodeType: LinkType): TreeNode[] {
+    const nodes = this.idsToNodes(famsIds.map(EntryId.fam), parentNode, childNodeType);
+    nodes.slice(1).forEach(node => node.primaryMarriage = nodes[0]);
+    return nodes;
+  }
+
+  private idsToNodes(entryIds: EntryId[], parentNode: TreeNode, childNodeType: LinkType, duplicateCheck = true): TreeNode[] {
+    return entryIds.map(entryId => this.idToNode(entryId, parentNode, childNodeType, duplicateCheck))
+      .filter(node => node != null);
+  }
+
+  private idToNode(entryId: EntryId, parentNode: TreeNode, childNodeType: LinkType, duplicateCheck = true): TreeNode {
+    if (this.isChildNodeTypeForbidden(childNodeType, parentNode)) return null;
+    const { id, isFam } = entryId;
+    if (isFam) {
+      const fam = this.data.getFam(id);
+      if (!fam || !fam.getFather() && !fam.getMother()) return null;  // Don't create fam nodes that are missing both husband and wife
+    }
+    const duplicateOf = this.queuedNodesById.get(id);
+    const node: TreeNode = {
+      id,
+      parentNode,
+      linkFromParentType: childNodeType,
+      childNodes: null,
+      linkStubs: null,
+      duplicateOf: duplicateCheck ? duplicateOf : null,
+    };
+    if (isFam) node.family = {id};
+    if (duplicateOf && duplicateCheck) duplicateOf.duplicated = true;
+    if (!duplicateOf) this.queuedNodesById.set(id, node);
+    return node;
+  }
+
+  private createLinkStubs(node: TreeNode): LinkType[] {
+    if (!node.duplicateOf && !node.duplicated && !node.primaryMarriage) return [];
+    const linkTypes = [LinkType.IndiParents, LinkType.IndiSiblings, LinkType.SpouseParents, LinkType.SpouseSiblings, LinkType.Children];
+    let arePresent = new Array(linkTypes.length).fill(false);
```

Comment:
This is an unusual way to represent a set. How about something like this:
```
const linkTypes = [];
if (this.isFamNode(node)) {
  const fam = this.data.getFam(node.family.id);
  if (fam) {
    const [indiParents, indiSiblings] =  this.areParentsAndSiblingsPresent(node.indi && node.indi.id);
    const [spouseParents, spouseSiblings] = this.areParentsAndSiblingsPresent(node.spouse && node.spouse.id);
    if (indiParents) linkTypes.push(LinkType.IndiParents);
    if (indiSiblings) linkTypes.push(LinkType.IndiSiblings);
    ...
```
or in a more functional way:
```
const linkTypes = [
  indiParents ? [LinkType.IndiParents] : [],
  indiSiblings ? [LinkType.IndiSiblings] : [],
  ...
].flat();
```